src/datanode/src/heartbeat/handler/open_region.rs

```diff
@@ -0,0 +1,200 @@
+            unreachable!("OpenRegionHandler: should be guarded by 'is_acceptable'");
+        };
+
+        ctx.finish();
+        let mailbox = ctx.mailbox.clone();
+        let self_ref = Arc::new(self.clone());
+
+        common_runtime::spawn_bg(async move {
+            let (engine, request) = OpenRegionHandler::prepare_request(region_ident);
+            let result = self_ref.open_region_inner(engine, request).await;
+            if let Err(e) = mailbox
+                .send((meta, OpenRegionHandler::map_result(result)))
+                .await
+            {
+                error!(e; "Failed to send reply to mailbox");
+            }
+        });
+        Ok(())
+    }
+}
+
+impl OpenRegionHandler {
+    fn map_result(result: Result<bool>) -> InstructionReply {
+        result.map_or_else(
+            |error| {
+                InstructionReply::OpenRegion(SimpleReply {
+                    result: false,
+                    error: Some(error.to_string()),
+                })
+            },
+            |result| {
+                InstructionReply::OpenRegion(SimpleReply {
+                    result,
+                    error: None,
+                })
+            },
+        )
+    }
+
+    fn prepare_request(ident: RegionIdent) -> (String, OpenTableRequest) {
+        let RegionIdent {
+            catalog,
+            schema,
+            table,
+            table_id,
+            region_number,
+            engine,
+        } = ident;
+
+        (
+            engine,
+            OpenTableRequest {
+                catalog_name: catalog,
+                schema_name: schema,
+                table_name: table,
+                table_id,
+                region_numbers: vec![region_number],
+            },
+        )
+    }
+
+    /// Returns true if table has been opened.
+    async fn check_table(
+        &self,
+        catalog_name: &str,
+        schema_name: &str,
+        table_name: &str,
+        region_numbers: &[RegionNumber],
+    ) -> Result<bool> {
+        if let Some(table) = self
+            .catalog_manager
+            .table(catalog_name, schema_name, table_name)
+            .await
+            .context(error::AccessCatalogSnafu)?
+        {
+            if region_numbers
+                .iter()
+                .map(|r| {
+                    table
+                        .contain_regions(*r)
+                        .with_context(|_| error::CheckRegionSnafu {
+                            table_name: format_full_table_name(
+                                catalog_name,
+                                schema_name,
+                                table_name,
+                            ),
+                            region_number: *r,
+                        })
+                })
+                .collect::<Result<Vec<_>>>()?
+                .into_iter()
+                .all(|r| r)
+            {
+                return Ok(true);
+            }
+        }
+        Ok(false)
+    }
```

Comment:
I'd suggest using for-loop instead iterators. First for-loop can early return when some region does not exist instead of iterates all regions to return. Second, you can log some context info when region does not exist instead of just return an `Ok(None)`.

```suggestion
        if let Some(table) = self
            .catalog_manager
            .table(catalog_name, schema_name, table_name)
            .await
            .context(error::AccessCatalogSnafu)?
        {
            for r in region_numbers {
                let region_exist =
                    table
                        .contain_regions(*r)
                        .with_context(|_| error::CheckRegionSnafu {
                            table_name: format_full_table_name(
                                catalog_name,
                                schema_name,
                                table_name,
                            ),
                            region_number: *r,
                        })?;
                if !region_exist {
                    warn!(
                        "Failed to check table:{}, region: {} does not exist",
                        format_full_table_name(catalog_name, schema_name, table_name,),
                        r
                    );
                    return Ok(false);
                }
            }
        }
        Ok(true)
```